src/Simplex_tree/include/gudhi/Simplex_tree.h

```diff
@@ -1469,6 +1472,124 @@ class Simplex_tree {
     }
   }
 
+  /** \brief Retrieve good values for extended persistence, and separate the 
+   * diagrams into the ordinary, relative, extended+ and extended- subdiagrams. 
+   * \post This function should be called only if extend_filtration has been called first!
+   * \post The coordinates of the persistence diagram points might be a little different than the
+   * original filtration values due to the internal transformation (scaling to [-2,-1]) that is 
+   * performed on these values during the computation of extended persistence.
+   * @param[in] dgm Persistence diagram obtained after calling this->extend_filtration 
+   * and this->get_persistence.
```

Comment:
Fixed it.

Reply:
(thinking aloud)
In a sense, what we would like to say is that this function is meant to take as argument the output of Persistent_cohomology_interface::get_persistence. Which would indicate that it might be a better fit somewhere in the python interface (possibly Simplex_tree_interface, I didn't think carefully).
However, Simplex_tree, as long as it provides extend, should also provide some way to get the original filtration values back. It (`decode_extended_filtration`?) could take as input either a Simplex_handle or a Filtration_value (and Extended_filtration_data if that appears), and return both the corresponding true filtration value, and whether it is UP or DOWN. Then the python wrapper would call that in a loop (it isn't clear if it should use the output of get_persistence, it might want to do its own loop like persistence_pairs or intervals_in_dimension do). And a C++ user could use this more basic function for other things than translating this very specific format of persistence diagram.

What do you think?

Reply:
Actually I modified the API a little bit so that there is no need to call `SimplexTree.persistence()` first. You can do it, but if you don't, it's run internally.

As for a "decoding" function, that would also be nice. Do you mean a function that would loop over all (original, i.e., not coned) simplices, and output the corresponding original filtration values?

Reply:
"decoding": let me rephrase what I said above, the basic functionality that we expect from Simplex_tree, after calling extend_filtration, is the ability to interpret the current Simplex_tree, and in particular its filtration values. That is, something like:
```c++
std::pair<Filtration_value,bool> decode_extended_filtration(Filtration_value f, Extended_filtration_data const& dat);
```
(or Simplex_handle as first argument, not sure)
where the output has both the original filtration value, and something that says if this was the value in the upward pass or the downward (aka relative) pass. In order to account for the apex, the bool may not be sufficient, it might need some `enum Extended_value_kind { UP, DOWN, EXTRA}` (where we would document that there is always exactly one EXTRA vertex, and that it is never paired in persistence). Or the bool might be enough if we document that in persistence, there is just one infinite interval, defined by an unmatched vertex, that this is not one of the original vertices but an extra one added for technical reasons, and that this function doesn't work for the filtration value of that vertex, so the user is responsible for not calling the function on it.
Ok, that's the basic functionality, and a C++ user should be able to make sense of the output of get_persistent_pairs or intervals_in_dimension (not persistent_betti_numbers, but let's ignore that).

The format `std::vector<std::pair<int, std::pair<Filtration_value, Filtration_value>>>` is something that basically only makes sense in the python interface because get_persistence returns that. In Simplex_tree, if we wanted to provide some function that acts on a whole diagram, it would be more natural if it took as input `vector<Persistent_interval>` for instance (the output of get_persistent_pairs). For that reason, I think we should move extended_persistence_subdiagrams to Simplex_tree_interface (or somewhere else in the python interface). And once it is there, on a similar level to get_persistence, you may not want to use the output of get_persistence and convert it, it may be nicer to work from get_persistent_pairs.

(Note #255 for the future, I hope that we won't have to build this strange list[tuple[int,tuple[float,float]]] forever)

I hope this was clearer than the previous message.

Reply:
Hmm, maybe there is one other element that is non-trivial to interpret for the C++ user: the dimension of features.

Reply:
Aaaah, I understand now. I like this idea. I've implemented the basic functionality in C++ and the python wrapper. Tell me if this is what you had in mind.